### PR TITLE
feat: add oauth2 standard connector

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

add oauth2 standard connector

## Testing

Tested locally, works fine.

### CODE REVIEW

**Review by wangsijie — COMMENTED** [review]
The logo is wrong?

File: packages/connector-oauth2/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+         * Maybe should make redirect_uri optional to be compatible to previous versions.
+         */
+        ...rest,
+        redirect_uri: configured_redirect_uri ?? redirectUri,
+      });
+
+      return `${parsedConfig.authorizationEndpoint}?${queryParameters.toString()}`;
+    }
+
+    // TODO: some flows do not access authorization endpoint, should fix API
```
Comment (wangsijie) [inline packages/connector-oauth2/src/index.ts:57]: Then better throw an error?

File: packages/connector-oauth2/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    } catch (error: unknown) {
+      if (error instanceof HTTPError) {
+        throw new ConnectorError(ConnectorErrorCodes.General, JSON.stringify(error.response.body));
+      }
+
+      throw error;
+    }
+  };
+
+const createGithubConnector: CreateConnector<SocialConnector> = async ({ getConfig }) => {
```
Comment (wangsijie) [inline packages/connector-oauth2/src/index.ts:131]: createOAuthConnector

File: packages/connector-oauth2/docs/config-template.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "oauthGrantType": "AuthorizationCode",
+  "authorizationEndpoint": "<authorization-endpoint>",
+  "tokenEndpoint": "<token-endpoint>",
+  "userInfoEndpoint": "<user-info-endpoint>",
+  "redirectUri": "<redirect-uri>",
+  "clientId": "<client-id>",
+  "clientSecret": "<client-secret>",
+  "scope": "<scope>"
```
Comment (wangsijie) [inline packages/connector-oauth2/docs/config-template.json:9]: Do we need to add "profileMapping" example?

File: packages/connector-oauth2/src/index.ts (hunk 1 above)
Comment (darcyYe) [reply packages/connector-oauth2/src/index.ts:57]: We call `getAuthorizationUri` and then `getUserInfo` for all social connectors, hence we can not simply throw an error here. Need to have at least two handlers when we get different authorization URIs.

File: packages/connector-oauth2/docs/config-template.json (hunk 3 above)
Comment (darcyYe) [reply packages/connector-oauth2/docs/config-template.json]: Added. Will elaborate on the configuration of different flows in README.md.

File: packages/connector-oauth2/README.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
+# OAuth standard connector
+
+The official Logto connector for OAuth 2.0 protocol.
+
+## Get started
+
+The OAuth connector enables Logto admin connector to an arbitrary social identity provider that supports OAuth 2.0 protocol.
```
Comment (wangsijie) [inline packages/connector-oauth2/README.md:7]: Please double check "Logto admin connector"

File: packages/connector-oauth2/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    const config = await getConfig(defaultMetadata.id);
+    validateConfig<OauthConfig>(config, oauthConfigGuard);
+    const parsedConfig = oauthConfigGuard.parse(config);
+
+    const parameterObject = snakecaseKeys(
+      omit(parsedConfig, ...oauthConfigGlobalKeys, 'clientSecret', 'grantType')
+    );
+
+    assert(
+      setSession,
```
Comment (wangsijie) [inline packages/connector-oauth2/src/index.ts:41]: Can we make "setSession" required in typing, and no need to check in the runtime?

**Review by wangsijie — APPROVED** [review]
Overall LGTM

File: packages/connector-oauth2/src/index.ts (hunk 5 above)
Comment (darcyYe) [reply packages/connector-oauth2/src/index.ts:41]: Good point, add LOG-4913 to track this. Let's address this in next cooldown cycle.

**Pull request merged**